docs/dm-config.rst
.. _dm-config:

全局配置
=====================

在阅读以下内容前，请确保您已经按照 :ref:`dm-installation` 正确部署了顶点云设备管理平台，并位于源码目录下（/path-to/zenith-monitor/）。

您可以根据您的环境修改源码目录下的 `config.py` 文件以配置服务器。

样例配置文件
----------------

您从 GitHub 仓库获取的源码中已经包含了一份默认的配置文件，这份配置文件中提供了一个基类 `Config` ，去掉注释后它的内容如下：

.. code-block:: python
   
   class Config:
    TEMP_PATH = 'temp'
    CLIENT_ADDRESS = '10.201.14.176'
    CLIENT_PORT = 50002
    SECRET_KEY = os.environ.get('SECRET_KEY') or \
                 '9d0e91f3372224b3ec7afec2' \
                 '4313e745efcf00ba4a5b767b' \
                 '35b17834d5f26efac197fd69' \
                 'd881dd92e629dbfdc2f1fbf6'
    SQLALCHEMY_COMMIT_ON_TEARDOWN = True
    SQLALCHEMY_TRACK_MODIFICATIONS = True
    ZENITH_MAIL_SUBJECT_PREFIX = '[顶点云设备管理]'
    ZENITH_MAIL_SENDER = os.environ.get('ZENITH_MAIL_SENDER') or \
                         'cloud-storage@forec.cn'
    ZENITH_TEMPFOLDER_LENGTH = 12
    ZENITH_INVALID_INFFIX = ['//', '\\', '/', '..', '%', '^', '&',
                             '*', '$', '!', '+', '#']
    EMAIL_ADMIN ='forec@bupt.edu.cn'
    ZENITH_RANDOM_PATH_ELEMENTS = ['a', 'b', 'c', 'd', 'e', 'f', 'g',
                                   'h', 'i', 'j', 'k', 'l', 'm', 'n',
                                   'o', 'p', 'q', 'r', 's', 't', 'u',
                                   'v', 'w', 'x', 'y', 'z', '1', '2',
                                   '3', '4', '5', '6', '7', '8', '9',
                                   '0', 'A', 'B', 'C', 'D', 'E', 'F',
                                   'G', 'H', 'I', 'J', 'K', 'L', 'M',
                                   'N', 'O', 'P', 'Q', 'R', 'S', 'T',
                                   'U', 'V', 'W', 'X', 'Y', 'Z']
    ZENITH_VALID_THUMBNAIL = ['.jpg', '.png', '.ico', '.jpeg']
    ZENITH_VALID_THUMBNAIL_SIZE = 512 * 1024

	
样例配置文件中，每项均有对应注释，您也可以查看下面的 :ref:`dm-config-detailed` 了解每一项的具体功能。

.. _dm-config-detailed:

样例文件详细解释
-------------------

配置文件中 `Config` 类的每一项对应配置如下：

1. ``TEMP_PATH`` ：服务器临时文件存放地址
2. ``CLIENT_ADDRESS`` ：模拟客户端 IP 地址/域名
3. ``CLIENT_PORT`` ：模拟客户端开放监听端口
4. ``SECRET_KEY`` ：用于服务器生成 token 使用的密钥，不可泄露，应当设置在服务器部署系统的环境变量中
5. ``SQLALCHEMY_COMMIT_ON_TEARDOWN`` ：为 ``True`` 时服务器终止运行时向数据库提交变动
6. ``SQLALCHEMY_TRACK_MODIFICATIONS`` ：为 ``True`` 时数据库将追踪服务器对数据的改动
7. ``ZENITH_MAIL_SUBJECT_PREFIX`` ：服务器向用户发送的验证邮件的主题前缀
8. ``ZENITH_MAIL_SENDER`` ：服务器向用户发送验证邮件使用的邮箱
9. ``ZENITH_TEMPFOLDER_LENGTH`` ：服务器在临时存储文件时生成的随机目录名长度
10. ``ZENITH_INVALID_INFFIX`` ：此列表中的字符不能出现在用户文件名中，否则视为不合法
11. ``ZENITH_RANDOM_PATH_ELEMENTS`` ：服务器生成的随机路径包含的元素
12. ``ZENITH_VALID_THUMBNAIL`` ：服务器允许用户上传的头像后缀名
13. ``ZENITH_VALID_THUMBNAIL_SIZE`` ：服务器允许用户上传的最大头像大小

`Config` 类仅仅是配置类的基类，你需要扩展此类才可完成配置。在 `config.py` 文件中，存在一些默认的 `Config` 子类，如 `LinuxConfig` 、 `WindowsConfig` 。

下面简单介绍 `LinuxConfig` 并以一个环境为例讲解如何配置。

`LinuxConfig` 类的内容如下，它是我（ `Forec`_ ）在我的云主机部署设备管理平台时使用的配置类：

.. code-block:: python
   
   class LinuxConfig(Config):
    ZENITH_SERVER_ADDRESS = 'cloud-monitor.forec.cn' # 服务器部署的域名/IP地址
    SERVER_NAME = ZENITH_SERVER_ADDRESS
    SQLALCHEMY_DATABASE_URI = 'sqlite:///' + os.path.join(basedir, 'work.db')
    MAIL_SERVER = 'smtp.exmail.qq.com'
    MAIL_PORT = 25 # SSL is 465
    MAIL_USE_TLS = True
    MAIL_USERNAME = "cloud-storage@forec.cn"
    MAIL_PASSWORD = os.environ.get('MAIL_PASSWORD')

可以看出， `LinuxConfig` 类添加了几个新的元素。新元素介绍如下：

1. ``ZENITH_SERVER_ADDRESS`` ：服务器部署使用的域名/IP地址
2. ``SERVER_NAME`` ：Flask 中的 *url_for* 函数使用的服务器名，通常保持和 *SERVER_NAME* 一致
3. ``SQLALCHEMY_DATABASE_URI`` ：服务器使用的数据库所在的路径
4. ``MAIL_SERVER`` ：服务器发送邮件使用的邮箱服务器
5. ``MAIL_PORT`` ：服务器使用 stmp 协议的端口号，通常为 25。使用 SSL 时设置为 465 但这取决于 *MAIL_SERVER* 是否支持 SSL
6. ``MAIL_USE_TLS`` ：是否启用安全连接发送邮件，通常设置为 ``True`` 
7. ``MAIL_USERNAME`` ：服务器发送邮件使用的邮箱帐号，通常和 *ZENITH_MAIL_SENDER* 保持一致
8. ``MAIL_PASSWORD`` ：服务器发送邮件使用的邮箱帐号的密码，通常保存在环境变量中

.. _dm-config-example:

下面通过一个实例环境解释如何配置。

例如，在安装 Ubuntu 16.04 的主机上部署顶点云设备管理平台，可参考的配置文件如下（使用扩展类）：

.. code-block:: python
   
   class MyConfig(Config):
    ZENITH_SERVER_ADDRESS = 'myaddress.my.io' # 自定义的域名，你需要先购买此域名并且映射到部署主机上
    SERVER_NAME = ZENITH_SERVER_ADDRESS
    SQLALCHEMY_DATABASE_URI = '/usr/local/cloud-monitor/mydb.db'
    	# 设置数据库为 /usr/local/cloud-monitor/mydb.db
    MAIL_SERVER = 'smtp.163.com'	# 使用 163 邮箱
    MAIL_PORT = 25
    MAIL_USE_TLS = True
    MAIL_USERNAME = "mycloud-monitor@163.com"
    MAIL_PASSWORD = os.environ.get('MAIL_PASSWORD') or '123456'
    	# 在环境变量中添加密码，若环境变量未找到对应值则使用 123456
    	
.. _dm-config-add:

添加自定义配置类到表
------------------------

在已经定义了自定义配置类后，你需要将自定义配置类添加到表驱动中以使 :ref:`dm-app-factory` 能根据我的配置类生成服务器实例。

在 `config.py` 中，有一个名为 `config` 的字典如下：

.. code-block:: python

   config = {
    'development' : DevelopmentConfig,      # 开发环境
    'linux': LinuxConfig,                   # 提供的 Linux 模板环境
    'windows': WindowsConfig,               # 提供的 Windows模板环境
    'testing' : TestingConfig,              # 测试环境
    'default' : DevelopmentConfig           # 默认为开发环境
    }
    
你需要添加自己的自定义配置类到此表中，如添加 ``'myconfig': MyConfig`` 。之后，修改 `manage.py` 中的第 21 行 ``app = create_app('default')`` 为 ``app = create_app('myconfig')`` 即可。
   
接下来请您阅读 :ref:`dm-quickstart` 。

.. _Forec: http://forec.cn


docs/_build/html/_sources/dm-config.rst.txt
.. _dm-config:

全局配置
=====================

在阅读以下内容前，请确保您已经按照 :ref:`dm-installation` 正确部署了顶点云设备管理平台，并位于源码目录下（/path-to/zenith-monitor/）。

您可以根据您的环境修改源码目录下的 `config.py` 文件以配置服务器。

样例配置文件
----------------

您从 GitHub 仓库获取的源码中已经包含了一份默认的配置文件，这份配置文件中提供了一个基类 `Config` ，去掉注释后它的内容如下：

.. code-block:: python
   
   class Config:
    TEMP_PATH = 'temp'
    CLIENT_ADDRESS = '10.201.14.176'
    CLIENT_PORT = 50002
    SECRET_KEY = os.environ.get('SECRET_KEY') or \
                 '9d0e91f3372224b3ec7afec2' \
                 '4313e745efcf00ba4a5b767b' \
                 '35b17834d5f26efac197fd69' \
                 'd881dd92e629dbfdc2f1fbf6'
    SQLALCHEMY_COMMIT_ON_TEARDOWN = True
    SQLALCHEMY_TRACK_MODIFICATIONS = True
    ZENITH_MAIL_SUBJECT_PREFIX = '[顶点云设备管理]'
    ZENITH_MAIL_SENDER = os.environ.get('ZENITH_MAIL_SENDER') or \
                         'cloud-storage@forec.cn'
    ZENITH_TEMPFOLDER_LENGTH = 12
    ZENITH_INVALID_INFFIX = ['//', '\\', '/', '..', '%', '^', '&',
                             '*', '$', '!', '+', '#']
    EMAIL_ADMIN ='forec@bupt.edu.cn'
    ZENITH_RANDOM_PATH_ELEMENTS = ['a', 'b', 'c', 'd', 'e', 'f', 'g',
                                   'h', 'i', 'j', 'k', 'l', 'm', 'n',
                                   'o', 'p', 'q', 'r', 's', 't', 'u',
                                   'v', 'w', 'x', 'y', 'z', '1', '2',
                                   '3', '4', '5', '6', '7', '8', '9',
                                   '0', 'A', 'B', 'C', 'D', 'E', 'F',
                                   'G', 'H', 'I', 'J', 'K', 'L', 'M',
                                   'N', 'O', 'P', 'Q', 'R', 'S', 'T',
                                   'U', 'V', 'W', 'X', 'Y', 'Z']
    ZENITH_VALID_THUMBNAIL = ['.jpg', '.png', '.ico', '.jpeg']
    ZENITH_VALID_THUMBNAIL_SIZE = 512 * 1024

	
样例配置文件中，每项均有对应注释，您也可以查看下面的 :ref:`dm-config-detailed` 了解每一项的具体功能。

.. _dm-config-detailed:

样例文件详细解释
-------------------

配置文件中 `Config` 类的每一项对应配置如下：

1. ``TEMP_PATH`` ：服务器临时文件存放地址
2. ``CLIENT_ADDRESS`` ：模拟客户端 IP 地址/域名
3. ``CLIENT_PORT`` ：模拟客户端开放监听端口
4. ``SECRET_KEY`` ：用于服务器生成 token 使用的密钥，不可泄露，应当设置在服务器部署系统的环境变量中
5. ``SQLALCHEMY_COMMIT_ON_TEARDOWN`` ：为 ``True`` 时服务器终止运行时向数据库提交变动
6. ``SQLALCHEMY_TRACK_MODIFICATIONS`` ：为 ``True`` 时数据库将追踪服务器对数据的改动
7. ``ZENITH_MAIL_SUBJECT_PREFIX`` ：服务器向用户发送的验证邮件的主题前缀
8. ``ZENITH_MAIL_SENDER`` ：服务器向用户发送验证邮件使用的邮箱
9. ``ZENITH_TEMPFOLDER_LENGTH`` ：服务器在临时存储文件时生成的随机目录名长度
10. ``ZENITH_INVALID_INFFIX`` ：此列表中的字符不能出现在用户文件名中，否则视为不合法
11. ``ZENITH_RANDOM_PATH_ELEMENTS`` ：服务器生成的随机路径包含的元素
12. ``ZENITH_VALID_THUMBNAIL`` ：服务器允许用户上传的头像后缀名
23. ``ZENITH_VALID_THUMBNAIL_SIZE`` ：服务器允许用户上传的最大头像大小

`Config` 类仅仅是配置类的基类，你需要扩展此类才可完成配置。在 `config.py` 文件中，存在一些默认的 `Config` 子类，如 `LinuxConfig` 、 `WindowsConfig` 。

下面简单介绍 `LinuxConfig` 并以一个环境为例讲解如何配置。

`LinuxConfig` 类的内容如下，它是我（ `Forec`_ ）在我的云主机部署设备管理平台时使用的配置类：

.. code-block:: python
   
   class LinuxConfig(Config):
    ZENITH_SERVER_ADDRESS = 'cloud-monitor.forec.cn' # 服务器部署的域名/IP地址
    SERVER_NAME = ZENITH_SERVER_ADDRESS
    SQLALCHEMY_DATABASE_URI = 'sqlite:///' + os.path.join(basedir, 'work.db')
    MAIL_SERVER = 'smtp.exmail.qq.com'
    MAIL_PORT = 25 # SSL is 465
    MAIL_USE_TLS = True
    MAIL_USERNAME = "cloud-storage@forec.cn"
    MAIL_PASSWORD = os.environ.get('MAIL_PASSWORD')

可以看出， `LinuxConfig` 类添加了几个新的元素。新元素介绍如下：

1. ``ZENITH_SERVER_ADDRESS`` ：服务器部署使用的域名/IP地址
2. ``SERVER_NAME`` ：Flask 中的 *url_for* 函数使用的服务器名，通常保持和 *SERVER_NAME* 一致
3. ``SQLALCHEMY_DATABASE_URI`` ：服务器使用的数据库所在的路径
4. ``MAIL_SERVER`` ：服务器发送邮件使用的邮箱服务器
5. ``MAIL_PORT`` ：服务器使用 stmp 协议的端口号，通常为 25。使用 SSL 时设置为 465 但这取决于 *MAIL_SERVER* 是否支持 SSL
6. ``MAIL_USE_TLS`` ：是否启用安全连接发送邮件，通常设置为 ``True`` 
7. ``MAIL_USERNAME`` ：服务器发送邮件使用的邮箱帐号，通常和 *ZENITH_MAIL_SENDER* 保持一致
8. ``MAIL_PASSWORD`` ：服务器发送邮件使用的邮箱帐号的密码，通常保存在环境变量中

.. _dm-config-example:

下面通过一个实例环境解释如何配置。

例如，在安装 Ubuntu 16.04 的主机上部署顶点云设备管理平台，可参考的配置文件如下（使用扩展类）：

.. code-block:: python
   
   class MyConfig(Config):
    ZENITH_SERVER_ADDRESS = 'myaddress.my.io' # 自定义的域名，你需要先购买此域名并且映射到部署主机上
    SERVER_NAME = ZENITH_SERVER_ADDRESS
    SQLALCHEMY_DATABASE_URI = '/usr/local/cloud-monitor/mydb.db'
    	# 设置数据库为 /usr/local/cloud-monitor/mydb.db
    MAIL_SERVER = 'smtp.163.com'	# 使用 163 邮箱
    MAIL_PORT = 25
    MAIL_USE_TLS = True
    MAIL_USERNAME = "mycloud-monitor@163.com"
    MAIL_PASSWORD = os.environ.get('MAIL_PASSWORD') or '123456'
    	# 在环境变量中添加密码，若环境变量未找到对应值则使用 123456
    	
.. _dm-config-add:

添加自定义配置类到表
------------------------

在已经定义了自定义配置类后，你需要将自定义配置类添加到表驱动中以使 :ref:`dm-app-factory` 能根据我的配置类生成服务器实例。

在 `config.py` 中，有一个名为 `config` 的字典如下：

.. code-block:: python

   config = {
    'development' : DevelopmentConfig,      # 开发环境
    'linux': LinuxConfig,                   # 提供的 Linux 模板环境
    'windows': WindowsConfig,               # 提供的 Windows模板环境
    'testing' : TestingConfig,              # 测试环境
    'default' : DevelopmentConfig           # 默认为开发环境
    }
    
你需要添加自己的自定义配置类到此表中，如添加 ``'myconfig': MyConfig`` 。之后，修改 `manage.py` 中的第 21 行 ``app = create_app('default')`` 为 ``app = create_app('myconfig')`` 即可。
   
接下来请您阅读 :ref:`dm-quickstart` 。

.. _Forec: http://forec.cn


config.py
# 作者：Forec
# 最后修改时间：2016-12-9
# 邮箱：forec@bupt.edu.cn
# 关于此文件: 服务器所有的配置信息在此文件中指定，具体功能由功能项后
#    的注释标明

import os
basedir = os.path.abspath(os.path.dirname(__file__))

class Config:
    #DEBUG = True
    TEMP_PATH = 'temp'
    CLIENT_ADDRESS = '127.0.0.1'#'10.201.14.162'#
    CLIENT_PORT = 50002
    SECRET_KEY = os.environ.get('SECRET_KEY') or \
                 '9d0e91f3372224b3ec7afec2' \
                 '4313e745efcf00ba4a5b767b' \
                 '35b17834d5f26efac197fd69' \
                 'd881dd92e629dbfdc2f1fbf6'         # 用于为安全操作生成 token 的密钥，不可泄露
    SQLALCHEMY_COMMIT_ON_TEARDOWN = True            # 数据库在服务器终止时 commit 变动
    SQLALCHEMY_TRACK_MODIFICATIONS = True           # 数据库追踪改动
    ZENITH_MAIL_SUBJECT_PREFIX = '[顶点云设备管理]'         # 服务器发送验证邮件的主题前缀
    ZENITH_MAIL_SENDER = os.environ.get('ZENITH_MAIL_SENDER') or \
                         'cloud-storage@forec.cn'   # 服务器向用户发送验证邮件的邮箱
    ZENITH_RANDOM_PATH_ELEMENTS = ['a', 'b', 'c', 'd', 'e', 'f', 'g',
                                       'h', 'i', 'j', 'k', 'l', 'm', 'n',
                                       'o', 'p', 'q', 'r', 's', 't', 'u',
                                       'v', 'w', 'x', 'y', 'z', '1', '2',
                                       '3', '4', '5', '6', '7', '8', '9',
                                       '0', 'A', 'B', 'C', 'D', 'E', 'F',
                                       'G', 'H', 'I', 'J', 'K', 'L', 'M',
                                       'N', 'O', 'P', 'Q', 'R', 'S', 'T',
                                       'U', 'V', 'W', 'X', 'Y', 'Z']
        # 用于生成随机路径的元素
    ZENITH_VALID_THUMBNAIL = ['.jpg', '.png', '.ico', '.jpeg']
        # 允许用户使用的头像后缀名
    ZENITH_VALID_THUMBNAIL_SIZE = 512 * 1024 # 最大 512K 头像
    ZENITH_TEMPFOLDER_LENGTH = 14


    @staticmethod
    def init_app(app):
        pass

class DevelopmentConfig(Config):            # 开发者环境配置
    SQLALCHEMY_DATABASE_URI = 'sqlite:///' + os.path.join(basedir, 'work.db')
    MAIL_SERVER = 'smtp.exmail.qq.com'
    MAIL_PORT = 25 # SSL is 465
    MAIL_USE_TLS = True
    MAIL_USERNAME = "cloud-storage@forec.cn"
    MAIL_PASSWORD = os.environ.get('MAIL_PASSWORD') or \
                    "Cloud-Storage-2016"

class TestingConfig(Config):                # 测试环境配置
    TESTING = True
    WTF_CSRF_ENABLED = False
    SQLALCHEMY_DATABASE_URI = 'sqlite:///' + os.path.join(basedir, 'temp/test.sqlite')
    ZENITH_SERVER_ADDRESS = '127.0.0.1'

class WindowsConfig(Config):
    #ZENITH_SERVER_ADDRESS = '127.0.0.1'# or 'cloud.forec.cn' # 服务器部署的域名/IP地址
    #SERVER_NAME = ZENITH_SERVER_ADDRESS
    SQLALCHEMY_DATABASE_URI = 'sqlite:///' + os.path.join(basedir, 'work.db')
    MAIL_SERVER = 'smtp.exmail.qq.com'
    MAIL_PORT = 25 # SSL is 465
    MAIL_USE_TLS = True
    MAIL_USERNAME = "cloud-storage@forec.cn"
    MAIL_PASSWORD = os.environ.get('MAIL_PASSWORD') or \
                    "Cloud-Storage-2016"

class LinuxConfig(Config):
    ZENITH_SERVER_ADDRESS = 'cloud-monitor.forec.cn' # 服务器部署的域名/IP地址
    SERVER_NAME = ZENITH_SERVER_ADDRESS
    SQLALCHEMY_DATABASE_URI = 'sqlite:///' + os.path.join(basedir, 'work.db')
    MAIL_SERVER = 'smtp.exmail.qq.com'
    MAIL_PORT = 25 # SSL is 465
    MAIL_USE_TLS = True
    MAIL_USERNAME = "cloud-storage@forec.cn"
    MAIL_PASSWORD = os.environ.get('MAIL_PASSWORD') or \
                    "Cloud-Storage-2016"

config = {
    'basis': Config,
    'development' : DevelopmentConfig,      # 开发环境
    'linux': LinuxConfig,                   # 提供的 Linux 模板环境
    'windows': WindowsConfig,               # 提供的 Windows模板环境
    'testing' : TestingConfig,              # 测试环境
    'default' : DevelopmentConfig               # 默认为开发环境
}
